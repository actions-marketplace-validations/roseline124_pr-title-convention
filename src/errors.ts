export enum ErrorType {
  TYPE_ERROR = 'TYPE_ERROR',
  SCOPE_ERROR = 'SCOPE_ERROR',
  SUBJECT_ERROR = 'SUBJECT_ERROR',
  SUBJECT_NOT_FOUND_ERROR = 'SUBJECT_NOT_FOUND_ERROR',
  UNKNOWN_ERROR = 'UNKNOWN_ERROR',
  SINGLE_COMMIT_ERROR = 'SINGLE_COMMIT_ERROR',
  // BRANCH_NAME_ERROR = 'BRANCH_NAME_ERROR', // @todo: implement later
}

type ErrorInfo = {
  errorWord?: string;
  availableWords?: string[];
  [field: string]: any;
};

export class ValidationError extends Error {
  constructor(message: string, public type = ErrorType.UNKNOWN_ERROR, public info?: ErrorInfo) {
    super(message);
  }
}
