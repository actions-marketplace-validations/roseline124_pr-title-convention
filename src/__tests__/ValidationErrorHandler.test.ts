import { ErrorType, ValidationError } from '../errors';
import { ValidationErrorHandler } from '../ValidationErrorHandler';

const mockClient = {
  async request(path: string, options?: Record<string, any>) {
    return options;
  },
};

const defaultPrContext = {
  base: { user: { login: 'tester' }, repo: { name: 'repo' }, number: 1 },
};

describe('ValidationErrorHandler', () => {
  const typeError = new ValidationError('type error', ErrorType.TYPE_ERROR, {
    errorWord: 'feet',
    availableWords: ['feat', 'fix'],
  });
  const scopeError = new ValidationError('scope error', ErrorType.SCOPE_ERROR, {
    errorWord: 'dug',
    availableWords: ['cat', 'dog'],
  });
  const subjectError = new ValidationError('subject error', ErrorType.SUBJECT_ERROR, {
    subject: ' Update the README with new information.',
    subjectPattern: '^(?!\\s?[A-Z]).+$',
  });

  describe("action is 'autofix'", () => {
    test('type이 틀리면 type을 수정한다.', async () => {
      const validationErrorHandler = new ValidationErrorHandler(mockClient as any, {
        ...defaultPrContext,
        title: 'feet: update the README with new information.',
      });
      const result: any = await validationErrorHandler.handleValidationError('autofix', [
        typeError,
      ]);

      expect(result.title).toBe('feat: update the README with new information.');
    });

    test('scope가 틀리면 scope를 수정한다.', async () => {
      const validationErrorHandler = new ValidationErrorHandler(mockClient as any, {
        ...defaultPrContext,
        title: 'feat(dug): update the README with new information.',
      });
      const result: any = await validationErrorHandler.handleValidationError('autofix', [
        scopeError,
      ]);

      expect(result.title).toBe('feat(dog): update the README with new information.');
    });
    test('type, scope 둘 다 틀리면 둘 다 수정한다.', async () => {
      const validationErrorHandler = new ValidationErrorHandler(mockClient as any, {
        ...defaultPrContext,
        title: 'feat(dug): update the README with new information.',
      });

      const result: any = await validationErrorHandler.handleValidationError('autofix', [
        typeError,
        scopeError,
      ]);

      expect(result.title).toBe('feat(dog): update the README with new information.');
    });
    test('subject가 틀리면 comment를 추가한다', async () => {
      const validationErrorHandler = new ValidationErrorHandler(mockClient as any, {
        ...defaultPrContext,
        title: 'feat: Update the README with new information.',
      });

      const result: any = await validationErrorHandler.handleValidationError('autofix', [
        subjectError,
      ]);

      expect(result.body).toContain('[PR Title Error Message]');
    });
  });

  describe('comment', () => {
    test('발생한 모든 에러의 메시지를 담아 comment를 추가한다', async () => {
      const validationErrorHandler = new ValidationErrorHandler(mockClient as any, {
        ...defaultPrContext,
        title: 'feet(dug): Update the README with new information.',
      });

      const result: any = await validationErrorHandler.handleValidationError('comment', [
        typeError,
        scopeError,
        subjectError,
      ]);

      expect(result.body).toMatchSnapshot();
    });
  });
});
