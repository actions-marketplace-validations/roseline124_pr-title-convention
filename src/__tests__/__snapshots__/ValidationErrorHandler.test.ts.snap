// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationErrorHandler comment 발생한 모든 에러의 메시지를 담아 comment를 추가한다 1`] = `
"💩 [PR Title Error Message] 
type error
scope error
subject error"
`;
